pokedex-with-bit/executors.py

```diff
@@ -0,0 +1,317 @@
+                 channel_axis: int = 1,
+                 on_gpu: bool = False,
+                 *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.channel_axis = channel_axis
+        self.model_path = model_path
+        self.on_gpu = on_gpu
+        self.logger = logger
+
+        if self.model_path and os.path.exists(self.model_path):
+            import tensorflow as tf
+            cpus = tf.config.experimental.list_physical_devices(
+                device_type='CPU')
+            gpus = tf.config.experimental.list_physical_devices(
+                device_type='GPU')
+            if self.on_gpu and len(gpus) > 0:
+                cpus.append(gpus[0])
+            tf.config.experimental.set_visible_devices(devices=cpus)
+            self.logger.info(f'BiT model path: {self.model_path}')
+            from tensorflow.python.keras.models import load_model
+            _model = load_model(self.model_path)
+            self.model = _model.signatures['serving_default']
+            self._get_input = tf.convert_to_tensor
+        else:
+            raise PretrainedModelFileDoesNotExist(f'model at {self.model_path} does not exist')
+
+    @requests
+    def encode(self, docs: DocumentArray, **kwargs) -> DocumentArray:
+        """
+        Encode data into a ndarray of `B x D`.
+        Where `B` is the batch size and `D` is the Dimension.
+
+        :param docs: an array in size `B`
+        :return: an ndarray in size `B x D`.
+        """
+        data = np.zeros((docs.__len__(),) + docs[0].blob.shape)
+        for index, doc in enumerate(docs):
+            data[index] = doc.blob
+        if self.channel_axis != -1:
+            data = np.moveaxis(data, self.channel_axis, -1)
+        _output = self.model(self._get_input(data.astype(np.float32)))
+        output = _output['output_1'].numpy()
+        for index, doc in enumerate(docs):
+            doc.embedding = output[index]
+        return docs
+
+
+class EmbeddingIndexer(Executor):
+    def __init__(self, index_file_name: str, **kwargs):
+        super().__init__(**kwargs)
+        self.index_file_name = index_file_name
+        if os.path.exists(self.save_path):
+            self._docs = DocumentArray.load(self.save_path)
+        else:
+            self._docs = DocumentArray()
+
+    @property
+    def save_path(self):
+        if not os.path.exists(self.workspace):
+            os.makedirs(self.workspace)
+        return os.path.join(self.workspace, self.index_file_name)
+
+    def close(self):
+        self._docs.save(self.save_path)
+
+    @requests(on='/index')
+    def index(self, docs: 'DocumentArray', **kwargs):
+        embedding_docs = DocumentArray()
+        for doc in docs:
+            embedding_docs.append(Document(id=doc.id, embedding=doc.embedding))
+        self._docs.extend(embedding_docs)
+
+    @requests(on='/search')
+    def search(self, docs: 'DocumentArray', parameters: Dict, **kwargs):
```

Comment:
please add type hint for return value

Reply:
https://github.com/jina-ai/examples/pull/626/commits/131bee3152cf88e330cc3b9856122a596ca984dd